File: leetcode/20_Valid_Parentheses.py
import unittest

class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        for ch in s:
            if ch in ("(", "[", "{"):
                stack.append(ch)
            elif len(stack) == 0:
                return False
            else:
                val = stack.pop()
                if not ((ch == ")" and val == "(") or (ch == "]" and val == "[") or (ch == "}" and val == "{")):
                    return False
        return len(stack) == 0

class TestValidParentheses(unittest.TestCase):
    def test_1(self):
        self.assertTrue(Solution().isValid("()"))    

    def test_2(self):
        self.assertTrue(Solution().isValid("()[]{}"))            

    def test_3(self):
        self.assertFalse(Solution().isValid("(]"))                    

    def test_4(self):
        self.assertFalse(Solution().isValid("([)]"))                    

    def test_5(self):
        self.assertTrue(Solution().isValid("{[]}"))                    

File: leetcode/20_Valid_Parentheses.rs
struct Solution {}

impl Solution {
    pub fn is_valid(s: String) -> bool {
        let mut stack = Vec::<char>::new(); 
        for ch in s.chars() {
            if ch == '(' || ch == '{' || ch == '[' {
                stack.push(ch);
            } else if stack.len() == 0 {
                return false;
            } else {
                let left = stack.pop().unwrap();
                if !((ch == ')' && left == '(') || (ch == ']' && left == '[') || (ch == '}' && left == '{')) {
                    return false;
                }
            }
        }
        stack.is_empty()
    }
}

#[cfg(test)]
mod valid_parentheses_test {
    use super::*;

    #[test]
    fn test_1() {
        assert!(Solution::is_valid("()".to_string()));
    }

    #[test]
    fn test_2() {
        assert!(Solution::is_valid("()[]{}".to_string()));
    }

    #[test]
    fn test_3() {
        assert!(!Solution::is_valid("(]".to_string()));
    }

    #[test]
    fn test_4() {
        assert!(!Solution::is_valid("([)]".to_string()));
    }

    #[test]
    fn test_5() {
        assert!(Solution::is_valid("{[]}".to_string()));
    }

}